src/coreclr/vm/jitinterface.cpp

```diff
@@ -11788,6 +11814,119 @@ void* CEEJitInfo::getFieldAddress(CORINFO_FIELD_HANDLE fieldHnd,
+            {
+                if (field->IsObjRef())
+                {
+                    OBJECTREF fieldObj = field->GetStaticOBJECTREF();
+                    if (fieldObj != NULL)
+                    {
+                        Object* obj = OBJECTREFToObject(fieldObj);
+                        if (GCHeapUtilities::GetGCHeap()->IsInFrozenSegment(obj))
+                        {
+                            *pValue = (uint64_t)obj;
+                            result = true;
+                        }
+                    }
+                    else
+                    {
+                        *pValue = 0;
+                        result = true;
+                    }
+                }
+                else
+                {
+                    void* fldAddr = field->GetStaticAddressHandle(field->IsRVA() ? nullptr : (void*)field->GetBase());
+                    if (fldAddr != nullptr)
+                    {
+                        _ASSERTE(!pEnclosingMT->ContainsGenericVariables());
+                        result = true;
+                        switch (field->GetFieldType())
+                        {
+                            case ELEMENT_TYPE_BOOLEAN:
+                                *pValue = (uint64_t)*((uint8_t*)fldAddr);
+                                break;
+                            case ELEMENT_TYPE_CHAR:
+                                *pValue = (uint64_t)*((uint16_t*)fldAddr);
+                                break;
+                            case ELEMENT_TYPE_I1:
+                                *pValue = (uint64_t)*((int8_t*)fldAddr);
+                                break;
+                            case ELEMENT_TYPE_U1:
+                                *pValue = (uint64_t)*((uint8_t*)fldAddr);
+                                break;
+                            case ELEMENT_TYPE_I2:
+                                *pValue = (uint64_t)*((int16_t*)fldAddr);
+                                break;
+                            case ELEMENT_TYPE_U2:
+                                *pValue = (uint64_t)*((uint16_t*)fldAddr);
+                                break;
+                            case ELEMENT_TYPE_I4:
+                                *pValue = (uint64_t)*((int32_t*)fldAddr);
+                                break;
+                            case ELEMENT_TYPE_U4:
+                                *pValue = (uint64_t)*((uint32_t*)fldAddr);
+                                break;
+                            case ELEMENT_TYPE_I8:
+                                *pValue = (uint64_t)*((int32_t*)fldAddr);
+                                break;
+                            case ELEMENT_TYPE_U8:
+                                *pValue = (uint64_t)*((uint64_t*)fldAddr);
+                                break;
+                            case ELEMENT_TYPE_R4:
+                                *pValue = (uint64_t)*((float*)fldAddr);
```

Comment:
This truncation looks odd.

Can we change this API to return bitwise data (e.g. take `char*` buffer and size and do `memcpy`)? With that shape it should potentially support other fields in the future too (e.g. structs).

Also, does this fix #53773? It seems like it does, but can you verify? We are seeing it a lot in `coreclr_tests.run` recently.

Reply:
> BTW: Storing frozen objects does not require write barrier. For example:
> 
> ```
> string[] a = new string[1];
> a[0] = "Hello"; // no GC write barrier needed if "Hello" is frozen object
> ```
> 
> Does this optimization happen already? If not, we just need to skip the write barrier for frozen icon handles, similar how we skip them for nulls.

Good point! Do you mind if I do it separately, I also wanted to e.g. check if it's worth optimizing pinning against frozen objects, etc. Separate commits also help to triage perf improvements/regressions

Reply:
To actually fix #53773 I think you can/should remove the part of `recGetFieldAddress/repGetFieldAddress` that records/replays the value.

Reply:
Fixed, thanks for noticing! It feels like there were spent quite a lot of efforts to make it work instead of introducing a simple jit-ee API 😄

Reply:
Yeah, I'm quite happy that we are finally fixing this.